zkevm-circuits/src/evm_circuit/execution/sstore.rs

```diff
@@ -467,636 +450,209 @@ impl<F: Field> SstoreTxRefundGadget<F> {
-            } else {
-                GasCost::SSTORE_RESET
-            }
-        } else {
-            GasCost::WARM_ACCESS
-        };
-        if is_warm {
-            warm_case_gas.as_u64()
+fn calc_expected_gas_cost(
+    value: eth_types::Word,
+    value_prev: eth_types::Word,
+    original_value: eth_types::Word,
+    is_warm: bool,
+) -> u64 {
+    let warm_case_gas = if value_prev == value {
+        GasCost::WARM_ACCESS
+    } else if original_value == value_prev {
+        if original_value == eth_types::Word::from(0) {
+            GasCost::SSTORE_SET
         } else {
-            warm_case_gas.as_u64() + GasCost::COLD_SLOAD.as_u64()
+            GasCost::SSTORE_RESET
         }
+    } else {
+        GasCost::WARM_ACCESS
+    };
+    if is_warm {
+        warm_case_gas.as_u64()
+    } else {
+        warm_case_gas.as_u64() + GasCost::COLD_SLOAD.as_u64()
     }
+}
 
-    fn calc_expected_tx_refund(
-        tx_refund_old: u64,
-        value: Word,
-        value_prev: Word,
-        committed_value: Word,
-    ) -> u64 {
-        let mut tx_refund_new = tx_refund_old;
-
-        if value_prev != value {
-            if committed_value == value_prev {
-                if (committed_value != Word::from(0)) && (value == Word::from(0)) {
-                    tx_refund_new += GasCost::SSTORE_CLEARS_SCHEDULE.as_u64();
-                }
+fn calc_expected_tx_refund(
+    tx_refund_old: u64,
+    value: eth_types::Word,
+    value_prev: eth_types::Word,
+    original_value: eth_types::Word,
+) -> u64 {
+    let mut tx_refund_new = tx_refund_old;
+
+    if value_prev != value {
+        if (original_value != value) && (value == eth_types::Word::from(0)) {
+            // CaseA
+            tx_refund_new += GasCost::SSTORE_CLEARS_SCHEDULE.as_u64();
+        }
+        if original_value == value {
+            if original_value != eth_types::Word::from(0) {
+                // CaseB
+                tx_refund_new += GasCost::SSTORE_RESET.as_u64() - GasCost::WARM_ACCESS.as_u64();
             } else {
-                if committed_value != Word::from(0) {
-                    if value_prev == Word::from(0) {
-                        tx_refund_new -= GasCost::SSTORE_CLEARS_SCHEDULE.as_u64()
-                    }
-                    if value == Word::from(0) {
-                        tx_refund_new += GasCost::SSTORE_CLEARS_SCHEDULE.as_u64()
-                    }
-                }
-                if committed_value == value {
-                    if committed_value == Word::from(0) {
-                        tx_refund_new +=
-                            GasCost::SSTORE_SET.as_u64() - GasCost::WARM_ACCESS.as_u64();
-                    } else {
-                        tx_refund_new +=
-                            GasCost::SSTORE_RESET.as_u64() - GasCost::WARM_ACCESS.as_u64();
-                    }
-                }
+                // CaseC
+                tx_refund_new += GasCost::SSTORE_SET.as_u64() - GasCost::WARM_ACCESS.as_u64();
```

Comment:
I think a `match` statement here will be much more clear than this.

Reply:
comments added. Main reason here is to follow original geth codes, so it is more easy for comparison / understanding

Reply:
Hmmm I understand. But at the same time, I think that if-else chains are really weird to read. While a `match` statement is always cristal clear.

Reply:
Also, Geth does use this syntax basically because the most similar thing to match they have is `switch` which works really bad in this situations.

Idk, not a must, if you think this is better not going to argue. As `match` usage might be just my preference.